### DISCUSSION TITLE
CustomHandlebarsExtensionFile not found

### DISCUSSION THREAD

**Discussion by egoux** [rfennell/AzurePipelines#1179] [Q&A]
Hi,

We are using the latest version of this extension. (3.77.5)
The task fails the first time reading an extended Handlebars function.
It looks like my extension file is not found.

Here are the specified parameter of the task

**

> customHandlebarsExtensionFile: '$(Build.SourcesDirectory)/Tools/ReleaseNotes/XplatGenerateReleaseNotes/EFS-RN-Tester-Handlebar-Extend.js'

**

However, the task works perfectly when I copy my file directly to the 3.77.5 extension directory, at the same level as GenerateReleaseNote.js for example.

Thanks to help me
Eric
(PS: My apologies for my English, more than average)

**Comment by rfennell** [comment]
I will need to check what is happening, but fully defining the path as you have done is normally correct.

If you run your pipeline with debug on (a variable `system.debug=true` or the diag checkbox on queuing the build) you should see something like the following wth the full path

```
Loading handlebars-helpers extension
##[debug]Attempting to load custom handlebars extension from D:\a\_tasks\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\3.77.5\customHandlebarsExtension.js}
##[debug]Custom handlebars extension file does not exist
Completed processing template
Writing output file D:\a\1\s\inline.md
```

So a couple of requests
1. Can you run a build with debug enabled and send me the log
2. Are you running on Windows, Mac or Linux (I have seem problems with path handling in the past)

**Reply by egoux** [reply]
Thanks for your answer.
Yes I'm running on Winddows.
I clone my pipeline and run it with debug on. 
The release notes task is always in Error.
It looks like it concatenates the directory where your extension is installed on the agent and the directory I specify for my Handlebars extension file.

![image](https://user-images.githubusercontent.com/94963486/148043338-6852cdfa-5993-4f61-88ad-07f7600c5b20.png)
[5.txt](https://github.com/rfennell/AzurePipelines/files/7806845/5.txt)

**Comment by rfennell** [comment]
The problem is that you have not specified the `customHandlebarsExtensionFolder`. If there is no `customHandlebarsExtensionFolder` value and you specify the `customHandlebarsExtensionFile` and leave the `customHandlebarsExtensionCode` empty, it defaults to use the folder the task is deployed to, exactly as you have seen.

So I think the settings you need, based on your log file are

```
customHandlebarsExtensionFile: EFS-RN-Tester-Handlebar-Extend.js
customHandlebarsExtensionFolder: $(Build.SourcesDirectory)/Tools/ReleaseNotes/XplatGenerateReleaseNotes
```

This structure is all a hang over from an earlier version of this task that only allowed extensions passed as an inline string `customHandlebarsExtensionCode`, where the `customHandlebarsExtensionFile` was only used as the name for the file (created in the task folder) to copy the inline string to. 

I did not want to break this past functionality, hence this somewhat confusing set of parameters

**Reply by egoux** [reply]
Ok, I understand, 
but I do this too, with these parameters in my task :

![image](https://user-images.githubusercontent.com/94963486/148087975-b4c7614a-1845-40a9-9fb9-10000907a5d4.png)

And the result is different but in error too.
I confirm that the extend source file is present in the source folder 

![image](https://user-images.githubusercontent.com/94963486/148088469-3ffe2653-769f-48f9-8475-4ed589d4b76b.png)

![image](https://user-images.githubusercontent.com/94963486/148088940-59a60dc1-4daf-4083-863e-2f98d3133f9d.png)

Strange, looking at the log, a closing brace has been added at the end of the file

![image](https://user-images.githubusercontent.com/94963486/148089906-3da7563b-41a9-4f50-9678-4cf46287f0b2.png)

**Reply by rfennell** [reply]
That is a known issue, See #832.  Whether you see it is dependant on what modules your custom extension references.  

If your custom extension has the line `const handlebars = require("handlebars");` at the start, and you are not in the same folder as the task `.js` file, then need to run NPM to make sure the required module is present e.g.

```
- task: Npm@1
  inputs:
    command: 'custom'
    workingDir: '$(System.DefaultWorkingDirectory)'
    customCommand: 'install handlebars'

- task: XplatGenerateReleaseNotes@3
  inputs:
    outputfile: '$(Build.ArtifactStagingDirectory)/releasenotes.md'
    templateLocation: 'File'
    templatefile: '$(System.DefaultWorkingDirectory)/template.hbs'
    customHandlebarsExtensionFile: 'custom.js'
    customHandlebarsExtensionFolder: '$(System.DefaultWorkingDirectory)'
```

Ths basic pattern is true for any included NPM modules

**Reply by egoux** [reply]
Oh sorry I hadn't seen this issue
Thank you for your prompt response.
Good evening.
Eric
